tests/unit/test_tools.py

```diff
@@ -1,11 +1,24 @@
 import numpy as np
 import unittest
 from elaston import tools
+from pint import UnitRegistry
+
+
+@tools.units(
+    inputs={"b": "angstrom", "x": "angstrom", "C": "GPa"}
```

Comment:
I'm not super familiar with `pint`, but I know I hate strings 😂 Is there a way/does it make sense to instead do something like `{"b": UnitRegistry.angstrom, ..., "C": UnitRegistry.GPa}`?

Reply:
Yeah that’s a bit of a problem with `pint`. You cannot derive a unit from the class. Instead you have to instantiate it, like

```python
ureg = UnitRegistry()
(1 * ureg.gigapascal * ureg.angstrom**3).to(ureg.electron_volt)
```

THE big problem is that if you instantiate it once, you have to use it consistently, i.e. this `ureg` must be somehow available to the user if it’s instantiated inside the class. I don’t think it’s impossible, but I find it super error-prone, so I tried to avoid it altogether by using strings instead.

Reply:
Maybe take the strings to be required to be attribute names of a unit registry?  Then in your decorator you can create your own unit registry for the life time of the function and `getattr` the relevant units.  Re: @liamhuber you could also allow users to give instances of pint units, but call `str` on them, then convert them back to units at call time using `ureg[...]` with the instance you create yourself.   This seems to work even for compound unit expressions, such `ureg.meter/ureg.seconds`.

Consider also abusing type annotations

```python
def stress(modulus: 'GPa', strain: 1) -> 'GPa':
  pass
```

Reply:
> Consider also abusing type annotations
> 
> ```python
> def stress(modulus: 'GPa', strain: 1) -> 'GPa':
>   pass
> ```
 I also thought about it. Can we do it just like this?

Reply:
> THE big problem is that if you instantiate it once, you have to use it consistently,

I haven't really worked with `pint` in depth before, I guess the idea is you instantiate `UnitRegistry` and then tell that instance what space of constants you're using and it works with it? Because my knee-jerk reaction was just to bypass what they're doing and wrap `UnitRegistry` as a `Singleton`. Now thinking about it (but not going as as to actually read anything), that would probably just shoot us in the foot.


> > Consider also abusing type annotations
> > ```python
> > def stress(modulus: 'GPa', strain: 1) -> 'GPa':
> >   pass
> > ```
> 
> I also thought about it. Can we do it just like this?

Here we might wind up bumping into `pyiron_workflow`, which is parsing the annotations and doesn't (currently) have any special handling for crazy annotation cases. At a minimum, I think you'd also have to explicitly get `typing.Annotations` involved in order to allow both type hinting _and_ unit hinting, e.g. `def stress(modulus: typing.Annotated[float, 'GPa'], strain: 1) -> typing.Annotated[float, 'GPa']:`. I'm pretty sure multiple annotations will break `pyiron_workflow` out of the box, but if we had a convention like the type hint always comes first, it should be easy enough to accommodate if you really want it.

Reply:
> Because my knee-jerk reaction was just to bypass what they're doing and wrap `UnitRegistry` as a `Singleton`. 

Yeah I also don’t really understand why it’s not already a `Singleton`. But for this the package seems to large so maybe I’m gonna live with what I got right now…